Return-Path: <linux-btrfs-owner@vger.kernel.org>
Received: from tux.applied-asynchrony.com ([unix socket])
	 by tux (Cyrus 2.5.17) with LMTPA;
	 Mon, 28 Aug 2023 10:08:12 +0200
X-Sieve: CMU Sieve 2.4
Received: from localhost (localhost [127.0.0.1])
	by tux.applied-asynchrony.com (Postfix) with ESMTP id 22532F01600
	for <holger@localhost>; Mon, 28 Aug 2023 10:08:12 +0200 (CEST)
Delivered-To: holger@applied-asynchrony.com
Received: from out1.vger.email (out1.vger.email [IPv6:2620:137:e000::1:20])
	by mail.itouring.de (Postfix) with ESMTP id C4D37CEF9CE
	for <holger@applied-asynchrony.com>; Mon, 28 Aug 2023 10:08:03 +0200 (CEST)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S230072AbjH1IHz (ORCPT <rfc822;holger@applied-asynchrony.com>);
        Mon, 28 Aug 2023 04:07:55 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:47202 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S230092AbjH1IHS (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Mon, 28 Aug 2023 04:07:18 -0400
Received: from dfw.source.kernel.org (dfw.source.kernel.org [139.178.84.217])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id AFA8DE7A
        for <linux-btrfs@vger.kernel.org>; Mon, 28 Aug 2023 01:06:49 -0700 (PDT)
Received: from smtp.kernel.org (relay.kernel.org [52.25.139.140])
        (using TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits)
         key-exchange X25519 server-signature RSA-PSS (2048 bits))
        (No client certificate requested)
        by dfw.source.kernel.org (Postfix) with ESMTPS id 13B9E63331
        for <linux-btrfs@vger.kernel.org>; Mon, 28 Aug 2023 08:06:49 +0000 (UTC)
Received: by smtp.kernel.org (Postfix) with ESMTPSA id F4189C433C8
        for <linux-btrfs@vger.kernel.org>; Mon, 28 Aug 2023 08:06:47 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=kernel.org;
        s=k20201202; t=1693210008;
        bh=Y9YSYRcRsvSy7pSXL0M07551lPqJpKEqIFHtMJz8kz8=;
        h=From:To:Subject:Date:In-Reply-To:References:From;
        b=pNmBHx83fH+sB2UuE1gUhTLRUjTL6ZI4Dlfye6Jp1HGiWIULBsh+hskJzfuaw3rP/
         yaUwPol03lCtNCDdXs7abm/8hkuSfKWMlU/kBlQBJCCVcXMYduf6VOjgYGCY1jh3DB
         7VmdydzMByAp34PMGLk7/GSaBkFOqDag0KvKAp8v/0u5Z8987hRA/D5F2yofLcbdB5
         x6w/KkNT7kfrnV1e4MbvzGSQpyWvCETulGtm0fQ7c3WMXaddfb71D8ToML51iB+F2T
         Lin5GERsINj/EOaNSLkvI96eJIPrOcOyIxFKbOPVkIBMgjacBzoRj84DzddIWw1qrl
         ZwU8FB651yK8w==
From:   fdmanana@kernel.org
To:     linux-btrfs@vger.kernel.org
Subject: [PATCH v2 1/3] btrfs: improve error message after failure to add delayed dir index item
Date:   Mon, 28 Aug 2023 09:06:42 +0100
Message-Id: <bb54e5910f8393a1404393138bc74df751d6655f.1693209858.git.fdmanana@suse.com>
X-Mailer: git-send-email 2.34.1
In-Reply-To: <cover.1693209858.git.fdmanana@suse.com>
References: <cover.1693209858.git.fdmanana@suse.com>
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Spam-Status: No, score=-2.1 required=5.0 tests=BAYES_00,DKIMWL_WL_HIGH,
        DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,DKIM_VALID_EF,
        RCVD_IN_DNSWL_BLOCKED,SPF_HELO_NONE,SPF_PASS autolearn=ham
        autolearn_force=no version=3.4.6
X-Spam-Checker-Version: SpamAssassin 3.4.6 (2021-04-09) on
        lindbergh.monkeyblade.net
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org
X-Text-Classification: accepted

From: Filipe Manana <fdmanana@suse.com>

If we fail to add a delayed dir index item because there's already another
item with the same index number, we print an error message (and then BUG).
However that message isn't very helpful to debug anything because we don't
know what's the index number and what are the values of index counters in
the inode and its delayed inode (index_cnt fields of struct btrfs_inode
and struct btrfs_delayed_node).

So update the error message to include the index number and counters.

We actually had a recent case where this issue was hit by a syzbot report
(see the link below).

Link: https://lore.kernel.org/linux-btrfs/00000000000036e1290603e097e0@google.com/
Signed-off-by: Filipe Manana <fdmanana@suse.com>
---
 fs/btrfs/delayed-inode.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/fs/btrfs/delayed-inode.c b/fs/btrfs/delayed-inode.c
index 08ecb4d0cc45..f9dae729811b 100644
--- a/fs/btrfs/delayed-inode.c
+++ b/fs/btrfs/delayed-inode.c
@@ -1498,9 +1498,10 @@ int btrfs_insert_delayed_dir_index(struct btrfs_trans_handle *trans,
 	ret = __btrfs_add_delayed_item(delayed_node, delayed_item);
 	if (unlikely(ret)) {
 		btrfs_err(trans->fs_info,
-			  "err add delayed dir index item(name: %.*s) into the insertion tree of the delayed node(root id: %llu, inode id: %llu, errno: %d)",
-			  name_len, name, delayed_node->root->root_key.objectid,
-			  delayed_node->inode_id, ret);
+"error adding delayed dir index item, name: %.*s, index: %llu, root: %llu, dir: %llu, dir->index_cnt: %llu, delayed_node->index_cnt: %llu, error: %d",
+			  name_len, name, index, btrfs_root_id(delayed_node->root),
+			  delayed_node->inode_id, dir->index_cnt,
+			  delayed_node->index_cnt, ret);
 		BUG();
 	}
 	mutex_unlock(&delayed_node->mutex);
-- 
2.40.1

